feat: add mdjs blog post (#4)
open-wc/blog-posts: 2020-04-introducing-mdjs-interactive-demos-everywhere/README.md

---
title: Introducing mdjs - interactive demos everywhere
published: false
description: Combine markdown and javascript to create the ultimate documentation to allow interactive demos everywhere.
tags: markdown, javascript, webcomponents, demos
cover_image: https://raw.githubusercontent.com/open-wc/blog-posts/feat/mdjs/2020-04-introducing-mdjs-interactive-demos-everywhere/images/aaron-burden-TNlHf4m4gpI-unsplash.jpg
---

Every shared code needs had written documentation to show what it can be used for and what the idea behind is.
Users should at least be able to get a high level understanding of what they are using, what they're using it for, and why.

On the web, we have many many different ways of writing documentation.
However, one thing almost all of them have in common is that they rely on [Markdown](https://en.wikipedia.org/wiki/Markdown) or some kind of variation of it.

And it's no surprise, because Markdown is supported practically everywhere (vscode, atom, github, gitlab, dev.to, npmjs, ...)

## For tools that do not run in the browser

> In that case, you will mostly share code snippets that people will need to run in their own projects so traditional static site generators like [Docusaurus](https://docusaurus.io/), [VuePress](https://vuepress.vuejs.org/), [Gatsby](https://www.gatsbyjs.org/), ... work great. All of them have full support for Markdown and allow you to easily create beautiful documentation pages with code snippets/highlighting, and more.

Reviewer: ```suggestion
In this case, you will mostly share code snippets that people will need to run in their own projects in which case traditional static site generators like [Docusaurus](https://docusaurus.io/), [VuePress](https://vuepress.vuejs.org/), [Gatsby](https://www.gatsbyjs.org/), et al work great. All of them have full support for Markdown and allow you to easily create beautiful documentation pages with code snippets/highlighting, and more.
```

And frankly, if that is your use case almost everything you need should be possible with those tools as long as you feel comfortable with the ecosystem/framework.

## For (visual) components that do run in the browser

In this context, users probably do expect a live demo to see all the different options of your component in action. So pure Markdown is usually not enough as we now want to actually execute code and "insert" our working component in our documentation somehow. This would require specialized handling for each framework.

### Vue

> For Vue you can use for example VuePress which auto registers all vue components in a certain folder and then you can use as normal html tags as Markdown allows html

Reviewer: ```suggestion
For Vue, as an example, you can use VuePress which auto registers all Vue components in a certain folder and then you can use as normal html tags since Markdown supports html
```

```
.
└─ .vuepress
  └─ components
      ├─ demo-1.vue
```

```html
<demo-1 />
```

- supports vue components and has "magical" import for them
- no support for generic javascript or passing properties to components

### React

For React you can use [Mdx](https://mdxjs.com/) which extends Markdown with JSX support. Mdx is available via multiple tools like [Gatsby](https://www.gatsbyjs.org/), [docz](https://www.docz.site/), [storybook](https://storybook.js.org/), etc.

```md
import { Chart } from '../components/chart'

# Here’s a chart

The chart is rendered inside our MDX document.
<Chart />
```

- supports import/export JavaScript
- passes everything through JSX
- Doesn't look _great_ on github, requires special tools in editors to get highlighting

## Limitations

What all these specialized tools have in common is that they require a specific build tooling setup to work.
For web components, none of that is actually needed. Markdown already allows for HTML. The only missing piece is how to load a web component through JavaScript?

## Introducing Markdown with JavaScript (mdjs)

The primary goals are

- minimal complexity
- follows progressive enhancement

> - Stick close to valid markdown syntax

Reviewer: ```suggestion
- stick close to valid markdown syntax
```

    import { html } from 'lit-html';
    ```
    # This is my component
    ```js story
    export const demo = () => html`
      <demo-wc-card header="HEADER"></demo-wc-card>
    `;
    ```

![mdjs story transform](https://raw.githubusercontent.com/open-wc/blog-posts/feat/mdjs/2020-04-introducing-mdjs-interactive-demos-everywhere/images/mdjs-story-transform-dev-to-size.gif)

> Link to [animation as slides](https://slides.com/dakmor/deck-mdjs#/)

Here another playground mimicking a full documentation page.

[==> Link to editable demo <==](https://webcomponents.dev/edit/PqrQkg3abvFJ7vxyZuqa?pm=1&sv=1)

## mdjs default docs page

Once all this meta-information is available you can render a specific docs page.

It basically comes down to generating this code which assigns the demo function to the actual web component.

```js
const stories = [{ key: 'demo', story: demo, code: demo }];
for (const story of stories) {
  const storyEl = rootNode.querySelector(`[mdjs-story-name="${story.key}"]`);
  storyEl.story = story.story;
  storyEl.code = story.code;
}
```

All of this happens under the hood for you 🤗

## Where can you use mdjs?

### You can use it locally via es-dev-server

> This will create a github like markdown view for all your local markdown files including live demos.

Reviewer: ```suggestion
Here i'll show you how you can create a github like markdown view for all your local markdown files including live demos.
```

![es-dev-server screenshot](https://raw.githubusercontent.com/open-wc/blog-posts/feat/mdjs/2020-04-introducing-mdjs-interactive-demos-everywhere/images/es-dev-server-screenshot.png)

> - Install dependency `npm i -D es-dev-server`

Reviewer: ```suggestion
- Install es-dev-server as a dependency by running `npm i -D es-dev-server`
```

> - Add to your `package.json`:

Reviewer: ```suggestion
- Add the following script to your `package.json`:
```

  ```json
  "scripts": {
    "start": "es-dev-server",
  }
  ```

- Create a `es-dev-server.config.js` in the root of your repo.

  ```js
  const { mdjsTransformer } = require('@mdjs/core');

  module.exports = {
    nodeResolve: true,
    open: 'README.md',
    watch: true,
    responseTransformers: [mdjsTransformer],
  };
  ```

> After executing `npm run start` you can happily browse you live documentation via [http://localhost:8000/README.md](http://localhost:8000/README.md).

Reviewer: ```suggestion
After executing `npm run start` you can happily browse your live documentation via [http://localhost:8000/README.md](http://localhost:8000/README.md).
```

You can see an example setup in the [demo-wc-card repo](https://github.com/daKmoR/demo-wc-card).

### You can use it via Storybook

If you want to work on individual components or get a list of all demos you can use Storybook.

![storybook screenshot](https://raw.githubusercontent.com/open-wc/blog-posts/feat/mdjs/2020-04-introducing-mdjs-interactive-demos-everywhere/images/storybook-screenshot.png)

- Install dependency `npm i -D @open-wc/demoing-storybook`

- Add to your `package.json`:

  ```json
  "scripts": {
    "storybook": "start-storybook --experimental-md-docs",
  }
  ```

- Adjust your `.storybook/main.js` to load markdown files

  ```js
  module.exports = {
    stories: ['../README.md', '../docs/**/*.md'],
    esDevServer: {
      nodeResolve: true,
      watch: true,
      open: true,
    },
  };
  ```

- Add to every markdown file that should be in storybook a name via

  ```js
  export default {
    title: 'My Group/My Awesome Component',
  };
  ```

> With that, you are good to go.

Reviewer: ```suggestion
And just like that, you are good to go.
```

> No additional changes to any files are needed a plugin will take care of it by converting your markdown files to the support mdx format of storybook.

Reviewer: ```suggestion
No additional changes to any files are needed; a plugin will take care of everything by converting your markdown files to the support Storybook's mdx format.
```

For more detailed information please see [https://open-wc.org/demoing-storybook/](https://open-wc.org/demoing-storybook/).

### Show it on github

Since Github supports markdown out of the box, we can take things even further by using Mdjs.

![github screenshot](https://raw.githubusercontent.com/open-wc/blog-posts/feat/mdjs/2020-04-introducing-mdjs-interactive-demos-everywhere/images/github-screenshot.png)

As it's not supported by github directly you will need a chrome extension called [mdjs-viewer](https://chrome.google.com/webstore/detail/mdjs-viewer/ifkkmomkjknligelmlcnakclabgohafe).

- Do you want to see a demo without opening a different page? mdjs-viewer!
- Do you want to show a live example of the issue you are having? mdjs-viewer!

==> INSERT SCREENSHOT/GIF/VIDEO <==

Almost looks like black magic, huh?
All you did was install a Chrome extension and suddenly Github got superpowers.

> All that is needed is to have Markdown files with the correct code fence blocks, and have your code be able to run from [unpkg.com](https://unpkg.com).

Reviewer: ```suggestion
All that you need is to have some Markdown files with the correct code fence blocks, and have your code up and running on [unpkg.com](https://unpkg.com).
```
